### PR TITLE
[chore] finalizer removal & teardown logic cleanup

### PR DESCRIPTION
**Pull request by shaneutt**
**What this PR does / why we need it**:

- corrects issues with finalizers that's been holding over since early railgun prototypes
- improves consistency on object deletions
- adds tests to prove new finalizer logic works appropriately

**Which issue this PR fixes**

fixes https://github.com/Kong/kubernetes-ingress-controller/issues/1515

### CODE REVIEW

**Comment by shaneutt** [comment]
flagging as do not merge because before we merge this I want to do some rebasing.

File: railgun/internal/proxy/proxy.go [hunk 1]
```diff
@@ -62,6 +62,9 @@ type Proxy interface {
 	// The delete action will asynchronously be converted to Kong DSL and applied to the Kong Admin API.
 	// A status will later be added to the object whether the configuration update succeeds or fails.
 	DeleteObject(obj client.Object) error
+
+	// ObjectExists indicates whether or not any version of the provided object is already present in the cache.
```
Comment (shaneutt) [inline railgun/internal/proxy/proxy.go:66]:
```suggestion
	// ObjectExists indicates whether or not any version of the provided object is already present in the proxy.
```

**Comment by rainest** [comment]
tl;dr
- This looks probably fine, and the changes other than the finalizer template code change itself look definitely fine.
- While this fixes any issues with Secrets by simply not attaching finalizers to them, I can't see why they were affected before but other resources weren't. Would that issue not also affect Ingresses, which still have finalizers?
- Do we think we need finalizers at all? After not really knowing about them, I've done research and discovered that they have ALL SORTS OF COMPLICATIONS. We didn't have them in 1.x and it doesn't appear that their use case is truly justified for us. Do we want to continue using them at all?

The commits not directly related to the finalizer stuff look fine, but I have _many_ questions about the finalizers.

Do we know why this was a problem for Secrets specifically? While this does remove our finalizer from them, it remains on other resources (e.g. Ingresses), and I'd expect that those also block namespace deletion if not removed--AFAIK the problem is that K8S won't allow deletion of namespaces without deleting its contents, and while it cascades these deletes automatically, the cascaded deletes won't finish until all finalizers are cleared from those objects.

The original section of the reconciler code in the current tip of next [did add our finalizer to everything](https://github.com/Kong/kubernetes-ingress-controller/blob/f42f1c926078e119466a743180b9c4a8fe6f1849/railgun/hack/generators/controllers/networking/main.go#L436-L444), but it also [apparently removed it from everything being deleted](https://github.com/Kong/kubernetes-ingress-controller/blob/f42f1c926078e119466a743180b9c4a8fe6f1849/railgun/hack/generators/controllers/networking/main.go#L418-L424). Was there some reason that Secrets failed timestamp condition there, or consistently resulted in an error when we called `err := r.Proxy.DeleteObject(obj)`? The latter seems more possible (it's our code _and_ it's more complex), but a brief review of that code indicates it's only deleting from cache, which should succeed.

That all said, this and the [error Jimin encountered](https://github.com/Kong/kubernetes-ingress-controller/pull/1487#issuecomment-876848214) (which is caused by me not properly adding permissions in #1457) makes me question whether we need the finalizers at all. The [kubebuilder example](https://book.kubebuilder.io/reference/using-finalizers.html) includes some logic we don't have to confirm the presence of the finalizer before doing things:

```
        if containsString(cronJob.GetFinalizers(), myFinalizerName) {
            // our finalizer is present, so lets handle any external dependency
            if err := r.deleteExternalResources(cronJob); err != nil {
```
but we don't: we simply [delete all deleting resources from our cache](https://github.com/Kong/kubernetes-ingress-controller/blob/f42f1c926078e119466a743180b9c4a8fe6f1849/railgun/hack/generators/controllers/networking/main.go#L418-L424) before stripping the finalizer, without any check to see if the finalizer is present. We still do that [in the proposed version](https://github.com/Kong/kubernetes-ingress-controller/blob/f76ff399bddf8baa5bbba5ec2eab7d16fc966375/railgun/hack/generators/controllers/networking/main.go#L434-L443), though with additional logic to requeue.

We won't remove our finalizer on resources that still include it until the proxy delete actually succeeds (because the requeue return precedes the finalizer clear), but we'll requeue indefinitely until the proxy delete does in fact succeed whether or not the finalizer is set.

The class annotation distinction between resources that get finalizers and those that don't seems rather arbitrary, as anything we reconcile influences Kong proxy configuration. For example, Ingresses (which need class annotations and will get a finalizer) result in routes, whereas KongPlugins attached to a Service (which do not need class annotations and will not get a finalizer) result in a plugin on a Kong service. It's not clear why we'd block resource deletion on one but not the other.

Kubernetes guidance on [when and why to use finalizers](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/) is unfortunately rather unclear, but my rough takeaway is that you use them when:
- You really want to avoid deleting something until some other condition is met (e.g. deleting a PV with an active PVC is almost certainly a bad idea).
- You really want to ensure some action elsewhere (e.g. failing to delete a cloud provider load balancer bound to a LoadBalancer means it sits around costing you money until you realize deletion failed and destroy it manually).

For us:
- We don't do anything like the first case currently, though we could do something like add a finalizer to a KongPlugin after we see an annotation referencing it on an Ingress: you would not be able to remove KongPlugins until you remove references to them.
- Our deletions are like the second case, though the justification is quite weak IMO. We do (now) try to eventually remove the object from the cache (which will in turn purge it from config) and in the worst case, can fix cases where retries never succeed by just restarting the controller, which will build a fresh cache from extant resources only. Furthermore, removing an object from the cache is _no guarantee that the config is not still active_, since it's quite possible to remove from cache, regenerate a config, and fail to apply config because the proxy rejects it.

In general, the UX for finalizers also seems poor:

1. Finalizers basically just hang kubectl (since the delete never finishes) or leave the resource in place until the finalizer clears (if you get bored and terminate the hung kubectl). It doesn't appear that there's any indication that the finalizer is what blocks it--there are no messages in `kubectl delete` output, even at V6, and no events in `kubectl describe` that refer to them)--beyond the presence of the finalizer in the YAML itself.
2. Finalizers don't provide information about deletion failures on resources themselves. You have to know to go check the finalizer owner's logs to see why it's not being removed, which is what we'd instruct users to do anyway if config is stuck (which usually happens because config is rejected, not because resources are stuck in cache).
3. It's furthermore not clear to me what should/does happen if you stop the controller and then want to remove something it finalized. It doesn't look like we remove the finalizer on shutdown, only on resource deletion. You're just stuck with those resources until you restart the controller.

File: railgun/Makefile [hunk 2]
```diff
@@ -166,18 +166,18 @@ test.integration: test.integration.dbless test.integration.postgres
 .PHONY: test.integration.dbless
 test.integration.dbless:
 	@./scripts/setup-integration-tests.sh
-	@TEST_DATABASE_MODE="off" GOFLAGS="-tags=integration_tests" go test -timeout 20m -race -v -count=1 -covermode=atomic -coverpkg=$(PKG_LIST) -coverprofile=$(COVERAGE_INTEGRATION_PROFILE) ./test/integration/
+	@TEST_DATABASE_MODE="off" GOFLAGS="-tags=integration_tests" go test -timeout 15m -race -v -count=1 -covermode=atomic -coverpkg=$(PKG_LIST) -coverprofile=$(COVERAGE_INTEGRATION_PROFILE) ./test/integration/
```
Comment (ccfishk) [inline railgun/Makefile:169]: Thought the worst case we tolerate is 30 minutes.

Comment (shaneutt) [reply railgun/Makefile:169]: I did some local testing with recent updates, and these tests average under 10 minutes to completion, so I figured I would give 5 minutes of buffer space but at this point, it would seem 20m is excessive based on new testing with recent testing environment updates. Let me know if that answers the question, or if you have further concerns, or if that satisfies your inquiry?

**Comment by shaneutt** [comment]
> tl;dr
> 
>     * This looks probably fine, and the changes other than the finalizer template code change itself look definitely fine.
> 
>     * While this fixes any issues with Secrets by simply not attaching finalizers to them, I can't see why they were affected before but other resources weren't. Would that issue not also affect Ingresses, which still have finalizers?

This was happening due to a race condition: the tests weren't _verifying_ that their namespaces were properly torn down, so the controller would be shut down before it could process all finalizers.

This is now fixed with the namespace teardown code.

>     * Do we think we need finalizers at all? After not really knowing about them, I've done research and discovered that they have ALL SORTS OF COMPLICATIONS. We didn't have them in 1.x and it doesn't appear that their use case is truly justified for us. Do we want to continue using them at all?
> 
> 
> The commits not directly related to the finalizer stuff look fine, but I have _many_ questions about the finalizers.
> 
> Do we know why this was a problem for Secrets specifically? While this does remove our finalizer from them, it remains on other resources (e.g. Ingresses), and I'd expect that those also block namespace deletion if not removed--AFAIK the problem is that K8S won't allow deletion of namespaces without deleting its contents, and while it cascades these deletes automatically, the cascaded deletes won't finish until all finalizers are cleared from those objects.
> 
> The original section of the reconciler code in the current tip of next [did add our finalizer to everything](https://github.com/Kong/kubernetes-ingress-controller/blob/f42f1c926078e119466a743180b9c4a8fe6f1849/railgun/hack/generators/controllers/networking/main.go#L436-L444), but it also [apparently removed it from everything being deleted](https://github.com/Kong/kubernetes-ingress-controller/blob/f42f1c926078e119466a743180b9c4a8fe6f1849/railgun/hack/generators/controllers/networking/main.go#L418-L424). Was there some reason that Secrets failed timestamp condition there, or consistently resulted in an error when we called `err := r.Proxy.DeleteObject(obj)`? The latter seems more possible (it's our code _and_ it's more complex), but a brief review of that code indicates it's only deleting from cache, which should succeed.
> 
> That all said, this and the [error Jimin encountered](https://github.com/Kong/kubernetes-ingress-controller/pull/1487#issuecomment-876848214) (which is caused by me not properly adding permissions in #1457) makes me question whether we need the finalizers at all. The [kubebuilder example](https://book.kubebuilder.io/reference/using-finalizers.html) includes some logic we don't have to confirm the presence of the finalizer before doing things:
> 
> ```
>         if containsString(cronJob.GetFinalizers(), myFinalizerName) {
>             // our finalizer is present, so lets handle any external dependency
>             if err := r.deleteExternalResources(cronJob); err != nil {
> ```
> 
> but we don't: we simply [delete all deleting resources from our cache](https://github.com/Kong/kubernetes-ingress-controller/blob/f42f1c926078e119466a743180b9c4a8fe6f1849/railgun/hack/generators/controllers/networking/main.go#L418-L424) before stripping the finalizer, without any check to see if the finalizer is present. We still do that [in the proposed version](https://github.com/Kong/kubernetes-ingress-controller/blob/f76ff399bddf8baa5bbba5ec2eab7d16fc966375/railgun/hack/generators/controllers/networking/main.go#L434-L443), though with additional logic to requeue.
> 
> We won't remove our finalizer on resources that still include it until the proxy delete actually succeeds (because the requeue return precedes the finalizer clear), but we'll requeue indefinitely until the proxy delete does in fact succeed whether or not the finalizer is set.
> 
> The class annotation distinction between resources that get finalizers and those that don't seems rather arbitrary, as anything we reconcile influences Kong proxy configuration. For example, Ingresses (which need class annotations and will get a finalizer) result in routes, whereas KongPlugins attached to a Service (which do not need class annotations and will not get a finalizer) result in a plugin on a Kong service. It's not clear why we'd block resource deletion on one but not the other.
> 
> Kubernetes guidance on [when and why to use finalizers](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/) is unfortunately rather unclear, but my rough takeaway is that you use them when:
> 
>     * You really want to avoid deleting something until some other condition is met (e.g. deleting a PV with an active PVC is almost certainly a bad idea).
> 
>     * You really want to ensure some action elsewhere (e.g. failing to delete a cloud provider load balancer bound to a LoadBalancer means it sits around costing you money until you realize deletion failed and destroy it manually).
> 
> 
> For us:
> 
>     * We don't do anything like the first case currently, though we could do something like add a finalizer to a KongPlugin after we see an annotation referencing it on an Ingress: you would not be able to remove KongPlugins until you remove references to them.
> 
>     * Our deletions are like the second case, though the justification is quite weak IMO. We do (now) try to eventually remove the object from the cache (which will in turn purge it from config) and in the worst case, can fix cases where retries never succeed by just restarting the controller, which will build a fresh cache from extant resources only. Furthermore, removing an object from the cache is _no guarantee that the config is not still active_, since it's quite possible to remove from cache, regenerate a config, and fail to apply config because the proxy rejects it.
> 
> 
> In general, the UX for finalizers also seems poor:
> 
>     1. Finalizers basically just hang kubectl (since the delete never finishes) or leave the resource in place until the finalizer clears (if you get bored and terminate the hung kubectl). It doesn't appear that there's any indication that the finalizer is what blocks it--there are no messages in `kubectl delete` output, even at V6, and no events in `kubectl describe` that refer to them)--beyond the presence of the finalizer in the YAML itself.
> 
>     2. Finalizers don't provide information about deletion failures on resources themselves. You have to know to go check the finalizer owner's logs to see why it's not being removed, which is what we'd instruct users to do anyway if config is stuck (which usually happens because config is rejected, not because resources are stuck in cache).
> 
>     3. It's furthermore not clear to me what should/does happen if you stop the controller and then want to remove something it finalized. It doesn't look like we remove the finalizer on shutdown, only on resource deletion. You're just stuck with those resources until you restart the controller.

I totally feel you on this.

Long story short? When this was originally written the intention was to ensure the relevant configurations in the proxy were _actually removed_ before letting the objects delete. Technical limitations regarding our integration with the Kong Admin API make this really hard to pull off right now so this didn't actually get implemented as originally intended.

So ultimately given the feasibility issues to make good use of them, I'll pull them all out, we can add them back later if we find solid use cases to do so. :+1:

**Comment by shaneutt** [comment]
@rainest check 1553b0fad3866dc9ba12b03ec109077c4c306f02 and let me know what you think :+1:

**Comment by shaneutt** [comment]
gonna get things working and then re-open :+1:

**Comment by shaneutt** [comment]
Fixed and working locally now :+1:

**Review by rainest — APPROVED** [review]

File: railgun/Makefile (hunk 2 above)
Comment (rainest) [reply railgun/Makefile:169]: Resolving to merge since this doesn't look like it should block. We can bump it up again if we're seeing timeouts that don't look like they're actually the result of a problem (some condition never completes because the condition isn't actually met, rather than just taking a long time because Github Actions slowness).
